common/settings.h

```diff
@@ -5,40 +5,81 @@
 #ifndef FLUTTER_COMMON_SETTINGS_H_
 #define FLUTTER_COMMON_SETTINGS_H_
 
+#include <fcntl.h>
 #include <stdint.h>
 
+#include <memory>
 #include <string>
 #include <vector>
 
+#include "lib/fxl/files/unique_fd.h"
+#include "lib/fxl/functional/closure.h"
+
 namespace blink {
 
+using TaskObserverAdd =
+    std::function<void(intptr_t /* key */, fxl::Closure /* callback */)>;
+using TaskObserverRemove = std::function<void(intptr_t /* key */)>;
+
 struct Settings {
-  bool enable_observatory = false;
-  // Port on target will be auto selected by the OS. A message will be printed
-  // on the target with the port after it has been selected.
-  uint32_t observatory_port = 0;
-  bool ipv6 = false;
-  bool start_paused = false;
-  bool trace_startup = false;
-  bool endless_trace_buffer = false;
-  bool enable_dart_profiling = false;
-  bool use_test_fonts = false;
-  bool dart_non_checked_mode = false;
-  bool enable_software_rendering = false;
-  bool using_blink = true;
-  std::string aot_shared_library_path;
+  // VM settings
+  std::string script_snapshot_path;
+  std::string kernel_snapshot_path;
+
   std::string aot_snapshot_path;
   std::string aot_vm_snapshot_data_filename;
   std::string aot_vm_snapshot_instr_filename;
   std::string aot_isolate_snapshot_data_filename;
   std::string aot_isolate_snapshot_instr_filename;
+
   std::string application_library_path;
+  std::string application_kernel_path;
+
+  std::string main_dart_file_path;
+  std::string packages_file_path;
+
   std::string temp_directory_path;
   std::vector<std::string> dart_flags;
+
+  // Isolate settings
+  bool start_paused = false;
+  bool trace_skia = false;
+  bool trace_startup = false;
+  bool endless_trace_buffer = false;
+  bool enable_dart_profiling = false;
+  bool dart_non_checked_mode = false;
+
+  // Observatory settings
+  bool enable_observatory = false;
+  // Port on target will be auto selected by the OS. A message will be printed
+  // on the target with the port after it has been selected.
+  uint32_t observatory_port = 0;
+  bool ipv6 = false;
+
+  // Font settings
+  bool use_test_fonts = false;
+
+  // Engine settings
+  TaskObserverAdd task_observer_add;
+  TaskObserverRemove task_observer_remove;
+  // The main isolate is current when this callback is made. This is a good spot
+  // to perform native Dart bindings for libraries not built in.
+  fxl::Closure root_isolate_create_callback;
+  // The isolate is not current and may have already been destroyed when this
+  // call is made.
+  fxl::Closure root_isolate_shutdown_callback;
+  bool enable_software_rendering = false;
+  bool using_blink = false;
+  bool skia_deterministic_rendering_on_cpu = false;
   std::string log_tag = "flutter";
+  std::string icu_data_path;
+
+  // Assets settings
+  int assets_dir = -1;
```

Comment:
Do you think it's worth suffixing with `_fd` for clarity? 

Based on grep, looks like this is only ever set for the fuchsia case. Looks like we do a fallback through this and `assets_path` where we actually use this, but in terms of actual usage, I'm assuming this is generally an either-or with assets_path? Do you think it's worth abstracting this behind a small class to reduce this to a single setting or not worth the bother?

Reply:
I really want to get rid of this field on the settings and put it on the run configuration. Right now, the settings object is used to configure the shell as well as the isolate in that shell (of which there can be many). Besides, there are also VM specific options (of which there can only be one in the process). Eventually, there should be no settings object and instead the embedder API ought to be flexible enough to allow configuring this on a per isolate basis. I would rather apply that in another patch.

Reply:
sgtm!